v3/lints/rfc/lint_key_usage_and_extended_key_usage_inconsistent.go

```diff
@@ -0,0 +1,201 @@
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/v3/lint"
+	"github.com/zmap/zlint/v3/util"
+)
+
+type KUAndEKUInconsistent struct{}
+
+func init() {
+	lint.RegisterLint(&lint.Lint{
+		Name:          "e_key_usage_and_extended_key_usage_inconsistent",
+		Description:   "The certificate MUST only be used for a purpose consistent with both key usage extension and extended key usage extension.",
+		Citation:      "RFC 5280, Section 4.2.1.12.",
+		Source:        lint.RFC5280,
+		EffectiveDate: util.RFC5280Date,
+		Lint:          NewKUAndEKUInconsistent,
+	})
+}
+
+func NewKUAndEKUInconsistent() lint.LintInterface {
+	return &KUAndEKUInconsistent{}
+}
+
+func (l *KUAndEKUInconsistent) Initialize() error {
+	return nil
+}
+
+// CheckApplies returns true when the certificate contains both a key usage
+// extension and an extended key usage extension.
+func (l *KUAndEKUInconsistent) CheckApplies(c *x509.Certificate) bool {
+	return util.IsSubscriberCert(c) && util.IsExtInCert(c, util.EkuSynOid) && util.IsExtInCert(c, util.KeyUsageOID)
+}
+
+// Execute returns an Error level lint.LintResult if the purposes of the certificate
+// being linted is not consistent with both extensions.
+func (l *KUAndEKUInconsistent) Execute(c *x509.Certificate) *lint.LintResult {
+	if len(c.ExtKeyUsage) > 1 {
+		return l.multiPurpose(c)
+	}
+	return l.strictPurpose(c)
+}
+
+// RFC 5280 4.2.1.12 on multiple purposes:
+//
+//	If multiple purposes are indicated the application need not recognize all purposes
+//	indicated, as long as the intended purpose is present.
+func (l *KUAndEKUInconsistent) multiPurpose(c *x509.Certificate) *lint.LintResult {
+	var mp = map[x509.KeyUsage]bool{}
+	for _, extKeyUsage := range c.ExtKeyUsage {
+		var i int
+		if _, ok := eku[extKeyUsage]; !ok {
+			return &lint.LintResult{Status: lint.Pass}
+		}
+		for ku := range eku[extKeyUsage] {
+			if i > 0 {
+				for mpku := range mp {
+					mp[mpku|ku] = true
+				}
+			}
+
+			mp[ku] = true
+			i++
+		}
+	}
+	if !mp[c.KeyUsage] {
+		return &lint.LintResult{
+			Status:  lint.Error,
+			Details: fmt.Sprintf("KeyUsage %v (%08b) inconsistent with multiple purpose ExtKeyUsage %v", util.GetKeyUsageStrings(c.KeyUsage), c.KeyUsage, util.GetEKUStrings(c.ExtKeyUsage)),
+		}
+	}
+	return &lint.LintResult{Status: lint.Pass}
+}
+
+// strictPurpose checks if the Key Usages (KU) included are permitted for each
+// indicated Extended Key Usage (EKU)
+func (l *KUAndEKUInconsistent) strictPurpose(c *x509.Certificate) *lint.LintResult {
+	for _, extKeyUsage := range c.ExtKeyUsage {
+		if _, ok := eku[extKeyUsage]; !ok {
+			continue
+		}
+		if !eku[extKeyUsage][c.KeyUsage] {
+			return &lint.LintResult{
+				Status:  lint.Error,
+				Details: fmt.Sprintf("KeyUsage %v (%08b) inconsistent with ExtKeyUsage %s", util.GetKeyUsageStrings(c.KeyUsage), c.KeyUsage, util.GetEKUString(extKeyUsage)),
+			}
+		}
+	}
+	return &lint.LintResult{Status: lint.Pass}
+}
+
+var eku = map[x509.ExtKeyUsage]map[x509.KeyUsage]bool{
```

Comment:
Thank you for setting up this truth table to be has declarative as possible.

I know that much has been said about the confusion that has been caused by these boolean expressions, so I appreciate the effort taken to make it clear,